pkg/render/egressgateway/egressgateway.go

```diff
@@ -0,0 +1,404 @@
+// limitations under the License.
+
+package egressgateway
+
+import (
+	"fmt"
+
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"strings"
+
+	operatorv1 "github.com/tigera/operator/api/v1"
+	"github.com/tigera/operator/pkg/components"
+	"github.com/tigera/operator/pkg/render"
+	rmeta "github.com/tigera/operator/pkg/render/common/meta"
+	"github.com/tigera/operator/pkg/render/common/podsecuritypolicy"
+	corev1 "k8s.io/api/core/v1"
+	policyv1beta1 "k8s.io/api/policy/v1beta1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	EGWPortName               = "health"
+	DefaultEGWVxlanPort int   = 4790
+	DefaultEGWVxlanVNI  int   = 4097
+	DefaultHealthPort   int32 = 8080
+)
+
+func EgressGateway(
+	config *Config,
+) render.Component {
+	return &component{
+		config: config,
+	}
+}
+
+type component struct {
+	config *Config
+}
+
+// Config contains all the config information EgressGateway needs to render component.
+type Config struct {
+	PullSecrets  []*corev1.Secret
+	Installation *operatorv1.InstallationSpec
+	OsType       rmeta.OSType
+	EgressGW     *operatorv1.EgressGateway
+
+	egwImage          string
+	EgressGWVxlanVNI  int
+	EgressGWVxlanPort int
+	Provider          operatorv1.Provider
+}
+
+func (c *component) ResolveImages(is *operatorv1.ImageSet) error {
+	reg := c.config.Installation.Registry
+	path := c.config.Installation.ImagePath
+	prefix := c.config.Installation.ImagePrefix
+
+	if c.config.OsType != c.SupportedOSType() {
+		return fmt.Errorf("Egress Gateway is supported only on %s", c.SupportedOSType())
+	}
+
+	var err error
+	c.config.egwImage, err = components.GetReference(components.ComponentEgressGateway, reg, path, prefix, is)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *component) SupportedOSType() rmeta.OSType {
+	return rmeta.OSTypeLinux
+}
+
+func (c *component) Objects() ([]client.Object, []client.Object) {
+	objects := []client.Object{}
+	if c.config.Provider == operatorv1.ProviderRKE2 {
+		objects = append(objects, c.egwPodSecurityPolicy())
+		objects = append(objects, c.egwClusterRole())
+		objects = append(objects, c.egwClusterRoleBinding())
+		objects = append(objects, c.egwServiceAccount())
+	}
+	objects = append(objects, c.egwDeployment())
+	return objects, nil
+}
+
+func (c *component) Ready() bool {
+	return true
+}
+
+func (c *component) egwDeployment() *appsv1.Deployment {
+	d := appsv1.Deployment{
+		TypeMeta: metav1.TypeMeta{Kind: "Deployment", APIVersion: "apps/v1"},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      c.config.EgressGW.Name,
+			Namespace: c.config.EgressGW.Namespace,
+			Labels:    c.config.EgressGW.Spec.Template.Metadata.Labels,
+		},
+		Spec: appsv1.DeploymentSpec{
+			Replicas: c.config.EgressGW.Spec.Replicas,
+			Selector: &metav1.LabelSelector{MatchLabels: map[string]string{"k8s-app": c.config.EgressGW.Name}},
```

Comment:
Does this still work if the user has specified `EgressGatewayMetadata.Labels` ?

Reply:
Yeah. It works fine even if we specify labels.